app/controlplane/internal/biz/attestationstate.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/google/uuid"
+	"golang.org/x/crypto/pbkdf2"
+	"google.golang.org/protobuf/proto"
+)
+
+type AttestationState struct {
+	State *v1.CraftingState
+}
+
+type AttestationStateRepo interface {
+	Initialized(ctx context.Context, workflowRunID uuid.UUID) (bool, error)
+	Save(ctx context.Context, workflowRunID uuid.UUID, state []byte) error
+	Read(ctx context.Context, workflowRunID uuid.UUID) ([]byte, error)
+	Reset(ctx context.Context, workflowRunID uuid.UUID) error
+}
+
+type AttestationStateUseCase struct {
+	repo      AttestationStateRepo
+	wfRunRepo WorkflowRunRepo
+}
+
+func NewAttestationStateUseCase(repo AttestationStateRepo, wfRunRepo WorkflowRunRepo) (*AttestationStateUseCase, error) {
+	return &AttestationStateUseCase{repo, wfRunRepo}, nil
+}
+
+func (uc *AttestationStateUseCase) Initialized(ctx context.Context, workflowID, runID string) (bool, error) {
+	run, err := uc.checkWorkflowRunInWorkflow(ctx, workflowID, runID)
+	if err != nil {
+		return false, fmt.Errorf("failed to check workflow run: %w", err)
+	}
+
+	initialized, err := uc.repo.Initialized(ctx, run.ID)
+	if err != nil {
+		return false, fmt.Errorf("failed to check initialized state: %w", err)
+	}
+
+	return initialized, nil
+}
+
+func (uc *AttestationStateUseCase) Save(ctx context.Context, workflowID, runID string, state *v1.CraftingState, passphrase string) error {
+	run, err := uc.checkWorkflowRunInWorkflow(ctx, workflowID, runID)
+	if err != nil {
+		return fmt.Errorf("failed to check workflow run: %w", err)
+	}
+
+	rawState, err := proto.Marshal(state)
+	if err != nil {
+		return fmt.Errorf("failed to marshal attestation state: %w", err)
+	}
+
+	encryptedState, err := encrypt(rawState, passphrase)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt attestation state: %w", err)
+	}
+
+	if err := uc.repo.Save(ctx, run.ID, encryptedState); err != nil {
+		return fmt.Errorf("failed to save attestation state: %w", err)
+	}
+
+	return nil
+}
+
+func (uc *AttestationStateUseCase) Read(ctx context.Context, workflowID, runID, passphrase string) (*AttestationState, error) {
+	run, err := uc.checkWorkflowRunInWorkflow(ctx, workflowID, runID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check workflow run: %w", err)
+	}
+
+	res, err := uc.repo.Read(ctx, run.ID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read attestation state: %w", err)
+	}
+
+	decryptedState, err := decrypt(res, passphrase)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt attestation state: %w", err)
+	}
+
+	state := &v1.CraftingState{}
+	if err := proto.Unmarshal(decryptedState, state); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal attestation state: %w", err)
+	}
+
+	return &AttestationState{State: state}, nil
+}
+
+func (uc *AttestationStateUseCase) Reset(ctx context.Context, workflowID, runID string) error {
+	run, err := uc.checkWorkflowRunInWorkflow(ctx, workflowID, runID)
```

Comment:
I am not sure about workflowID in this method. I was confused about why we need it here , but I don't have any better suggestions at the moment :)

Some questions I asked myself?
- If we want to check whether the runID belongs to workflowId, should we do it on the service level?
- The checkWorkflowRunInWorkflow method looks to me more like a helper to find a wfRun object, and workflowId is just something extra there. Should we rename it?
- Wouldn't using an encryption key suffice in preventing any attempts of spoofing? I know we use the robot account today, and it might be shared

Reply:
> If we want to check whether the runID belongs to workflowId, should we do it on the service level?

we definitely could, but I am trying to make sure this logic is also easily testable, and our `biz` layer is easy to test both unitarily and functionally. Another reason to add it here is to keep the service layer to just do request handling.

> The checkWorkflowRunInWorkflow method looks to me more like a helper to find a wfRun object, and workflowId is just something extra there. Should we rename it?

That's true, the fact that it returns a run is confusing. Will change it. Thanks!

> Wouldn't using an encryption key suffice in preventing any attempts of spoofing? I know we use the robot account today, and it might be shared

Could you elaborate on how this would work?